### ISSUE TITLE
C2482 with MSVC 19.0.24215.1

### ISSUE THREAD

**Issue by Kentzo** [gabime/spdlog#631]
The compiler fails on [static thread_local const size_t tid = _thread_id()](https://github.com/gabime/spdlog/blob/v0.16.3/include/spdlog/details/os.h#L358):

```
details/os.h(358): error C2482: 'tid': dynamic initialization of thread local data not allowed in WinRT code
```

**Comment by gabime** [comment]
Hi, on which platfom and compiler versions?

**Comment by Kentzo** [comment]
Windows 8.1, MSVC 19.0.24215.1 (I believe it’s msvs 2015 update 3).

**Comment by gabime** [comment]
yes,  but why does it say “WinRT code”?

**Comment by Kentzo** [comment]
That's a good question. Here are the flags passed to the compiler:

```
[Step 7/9] Building CXX object example.cpp.obj
[Step 7/9] Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24215.1 for x64
[Step 7/9] Copyright (C) Microsoft Corporation.  All rights reserved.
[Step 7/9] 
[Step 7/9] cl   /TP -DHAVE_MS_C_RUNTIME -DHAVE_WIN32_IPC -DHAVE_WIN32_PROC -DHAVE_WIN32_THREADS -DHAVE_WINSOCK -DNOMINMAX -DSPDLOG_WCHAR_FILENAMES=1 -DWIN32=1 -DWIN32_LEAN_AND_MEAN -D_CRT_NONSTDC_NO_DEPRECATE=1 -D_CRT_SECURE_NO_WARNINGS=1 -D_WIN32=1 -D_WINSOCK_DEPRECATED_NO_WARNINGS=1 -D__thread=__declspec(thread) /DWIN32 /D_WINDOWS /W3 /GR /EHsc -MP /MD /Zi /O2 /Ob1 /DNDEBUG -Ox -Oy -Oi -GL -Gw -Gy -fp:fast   -wd4267 -wd4244 -wd4800 /ZW /EHsc "/AIC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\\vcpackages" /AI$(VCInstallDir)\vcpackages /AI$(LibraryWPath) /Foexample.cpp.obj /Fdexample.dir\ /FS -c example.cpp
```

**Comment by Kentzo** [comment]
Perhaps the #ifdef needs to be augumented to check `WINAPI_FAMILY_APP`? Although I do not know whether this check should be bound to the compiler.

**Comment by gabime** [comment]
According to https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2482 it should be only on 2013 or earlier. Are you sure it msvc 2015?

**Comment by Kentzo** [comment]
Compiler advertises itself as `Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24215.1 for x64`.

**Comment by gabime** [comment]
So, the mistery remains why is it targeting WinRT..

**Comment by Kentzo** [comment]
> So, the mistery remains why is it targeting WinRT.

It's not a mystery: we specifically enabled WinRT to use geolocation API which is not available otherwise.

> Any way you can disable this in tweakme.h (SPDLOG_DISABLE_TID_CACHING)

That's how I solved it for now.

> Could you try checking the “__cplusplus_winrt” in the ifdef too? and post a pull if it helps?

Sure.

**Comment by wblume** [comment]
I also have this problem.  I'm compiling for UniversalWindowsPlatform using Visual Studio Community 2017 (Version 15.4.0).

   I used SPDLOG_DISABLE_TID_CACHING to work around

**Comment by gabime** [comment]
Fixed in 80163dc6c1d3e31fa26c8b9515ecfbf76767442f
